indico/migrations/versions/20191023_1820_2496c4adc7e9_add_track_principal.py

```diff
@@ -0,0 +1,304 @@
+        (reviewers.user_id = conveners.user_id AND
+        reviewers.track_id = conveners.track_id)))
+    '''
+
+    for user_id, track_id in conn.execute(track_conveners):
+        _create_track_acl_entry(conn, user_id, track_id, ['convene'])
+
+    # Update event acl entries
+
+    event_reviewers = '''
+        SELECT reviewers.user_id, reviewers.event_id, principals.permissions
+        FROM events.track_abstract_reviewers as reviewers
+        JOIN events.principals as principals
+        ON reviewers.user_id = principals.user_id AND reviewers.event_id = principals.event_id
+        WHERE reviewers.event_id IS NOT NULL
+    '''
+
+    for user_id, event_id, permissions in conn.execute(event_reviewers):
+        bisect.insort(permissions, 'review_all_abstracts')
+        _update_event_acl_entry(conn, user_id, event_id, permissions)
+
+    event_conveners = '''
+        SELECT conveners.user_id, conveners.event_id, principals.permissions
+        FROM events.track_conveners as conveners
+        JOIN events.principals as principals
+        ON conveners.user_id = principals.user_id AND conveners.event_id = principals.event_id
+        WHERE conveners.event_id IS NOT NULL
+    '''
+
+    for user_id, event_id, permissions in conn.execute(event_conveners):
+        bisect.insort(permissions, 'convene_all_abstracts')
+        _update_event_acl_entry(conn, user_id, event_id, permissions)
+
+
+def upgrade():
+    op.create_table(
+        'track_principals',
+        sa.Column('read_access', sa.Boolean(), nullable=False),
+        sa.Column('full_access', sa.Boolean(), nullable=False),
+        sa.Column('permissions', postgresql.ARRAY(sa.String()), nullable=False),
+        sa.Column('id', sa.Integer(), nullable=False),
+        sa.Column('track_id', sa.Integer(), nullable=False, index=True),
+        sa.Column('local_group_id', sa.Integer(), nullable=True, index=True),
+        sa.Column('mp_group_provider', sa.String(), nullable=True),
+        sa.Column('event_role_id', sa.Integer(), nullable=True, index=True),
+        sa.Column('mp_group_name', sa.String(), nullable=True),
+        sa.Column('user_id', sa.Integer(), nullable=True, index=True),
+        sa.Column(
+            'type',
+            PyIntEnum(PrincipalType, exclude_values={PrincipalType.network}),
+            nullable=False,
+        ),
+        sa.Column('email', sa.String(), nullable=True, index=True),
+        sa.CheckConstraint(
+            u'NOT full_access', name=op.f('ck_track_principals_no_full_access')
+        ),
+        sa.CheckConstraint(
+            u'NOT read_access', name=op.f('ck_track_principals_no_read_access')
+        ),
+        sa.CheckConstraint(
+            u'email IS NULL OR email = lower(email)',
+            name=op.f('ck_track_principals_lowercase_email'),
+        ),
+        sa.CheckConstraint(
+            u'read_access OR full_access OR array_length(permissions, 1) IS NOT NULL',
```

Comment:
will `array_length` ever return `NULL`? I think it should always return an `int`

Reply:
Ahh, I see now.. docs are a bit misleading :P (it returns `NULL` for an empty array and docs specify that it always returns an `int`)